File: src/core/karana.ts
import { normalize360 } from '../utils/angle';
import { KARANA_SPAN } from '../utils/constants';
import type { KaranaInfo } from '../types/elements';

/**
 * Karana = half-Tithi = 6° of Moon-Sun elongation. 60 Karanas per lunar month.
 *
 * Mapping:
 *   index 0:       Kimstughna (fixed)
 *   index 1–56:    7-Karana movable cycle (Bava→Vishti, repeating 8 times)
 *   index 57–59:   Shakuni, Chatushpada, Naga (fixed)
 */
export function computeKaranaFromLongitudes(
  siderealMoon: number,
  siderealSun: number,
  name: string,
): KaranaInfo {
  const angle = normalize360(siderealMoon - siderealSun);
  const index = Math.floor(angle / KARANA_SPAN);
  const elapsed = angle - index * KARANA_SPAN;
  const completionPercentage = (elapsed / KARANA_SPAN) * 100;
  return {
    index,
    name,
    type: getKaranaType(index),
    completionPercentage: Math.round(completionPercentage * 100) / 100,
    endTime: null,
  };
}

export function getKaranaType(index: number): 'fixed' | 'movable' {
  return index === 0 || index >= 57 ? 'fixed' : 'movable';
}

export function getKaranaIndexAtTime(
  date: Date,
  getCachedMoon: (d: Date) => number,
  getCachedSun: (d: Date) => number,
): number {
  const angle = normalize360(getCachedMoon(date) - getCachedSun(date));
  return Math.floor(angle / KARANA_SPAN);
}

/** Direct longitude → index. Used by the orchestrator at sunrise. */
export function getKaranaIndex(siderealMoon: number, siderealSun: number): number {
  return Math.floor(normalize360(siderealMoon - siderealSun) / KARANA_SPAN);
}

File: source/go/internal/core/karana.go
package core

import (
	"math"

	"github.com/ishankgupta95/panchang/source/go/v5/internal/jsnum"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/types"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/utils"
)

func ComputeKaranaFromLongitudes(siderealMoon, siderealSun float64, name string) types.KaranaInfo {
	angle := utils.Normalize360(siderealMoon - siderealSun)
	index := int(math.Floor(angle / utils.KaranaSpan))
	elapsed := angle - float64(float64(index)*utils.KaranaSpan) // anti-FMA barrier
	completionPercentage := (elapsed / utils.KaranaSpan) * 100
	return types.KaranaInfo{
		Index:                index,
		Name:                 name,
		Type:                 GetKaranaType(index),
		CompletionPercentage: jsnum.Round(completionPercentage*100) / 100,
		EndTime:              nil,
	}
}

func GetKaranaType(index int) types.KaranaType {
	if index == 0 || index >= 57 {
		return types.KaranaFixed
	}
	return types.KaranaMovable
}

func GetKaranaIndexAtTime(ms int64, getCachedMoon, getCachedSun LongitudeAt) int {
	angle := utils.Normalize360(getCachedMoon(ms) - getCachedSun(ms))
	return int(math.Floor(angle / utils.KaranaSpan))
}

func GetKaranaIndex(siderealMoon, siderealSun float64) int {
	return int(math.Floor(utils.Normalize360(siderealMoon-siderealSun) / utils.KaranaSpan))
}
